traits/tests/test_traits_listener.py

```diff
@@ -0,0 +1,468 @@
+# (C) Copyright 2005-2020 Enthought, Inc., Austin, TX
+# All rights reserved.
+#
+# This software is provided without warranty under the terms of the BSD
+# license included in LICENSE.txt and may be redistributed only under
+# the conditions described in the aforementioned license. The license
+# is also available online at http://www.enthought.com/licenses/BSD.txt
+#
+# Thanks for using Enthought open source!
+
+"""
+Unittest tests on ListenerParser, ListenerBase etc.
+See test_listeners for integration tests on the public API using these objects.
+
+"""
+from functools import partial
+import unittest
+
+from traits import traits_listener
+from traits.api import (
+    TraitError,
+    pop_exception_handler,
+    push_exception_handler,
+)
+
+
+def assert_listener_item_equal(test_case, item1, item2, msg=None):
+    """ Assertion function for comparing two instances of ListenerItem.
+    """
+
+    def get_msg(name, msg):
+        return "{name} mismatched. {msg}".format(
+            name=name,
+            msg="" if msg is None else msg
+        )
+
+    test_case.assertEqual(
+        item1.name, item2.name,
+        msg=get_msg("name", msg),
+    )
+    test_case.assertEqual(
+        item1.metadata_name, item2.metadata_name,
+        msg=get_msg("metadata_name", msg),
+    )
+    test_case.assertEqual(
+        item1.metadata_defined, item2.metadata_defined,
+        msg=get_msg("metadata_defined", msg),
+    )
+    test_case.assertEqual(
+        item1.is_any_trait, item2.is_any_trait,
+        msg=get_msg("is_any_trait", msg),
+    )
+    test_case.assertEqual(
+        item1.dispatch, item2.dispatch,
+        msg=get_msg("dispatch", msg),
+    )
+    test_case.assertEqual(
+        item1.notify, item2.notify,
+        msg=get_msg("notify", msg),
+    )
+    test_case.assertEqual(
+        item1.is_list_handler, item2.is_list_handler,
+        msg=get_msg("is_list_handler", msg),
+    )
+    test_case.assertEqual(
+        item1.type, item2.type,
+        msg=get_msg("type", msg),
+    )
+    if item1.next is item2.next:
+        # avoid recursion
+        pass
+    else:
+        test_case.assertEqual(
+            item1.next, item2.next,
+            msg=get_msg("next", msg),
+        )
```

Comment:
I could use `trait_names` here and deal with `next` separately.

Reply:
Could do, but I think it's fine as is. The advantage of the repetition is that the traceback for a failed test will tell you more about which part failed. If all those `assertEqual`s are combined, we'll have to be more careful about constructing the failure message to say exactly which part failed.

Reply:
Also, I'd be quite happy if it turned out to be possible to make `ListenerItem` *not* a `HasTraits` subclass at some point in the future.

Reply:
Ooops, I sneakily made this last minute change - will revert to what ever you approved :D